compiler/qsc_data_structures/src/language_features.rs

```diff
@@ -0,0 +1,76 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+use clap::ValueEnum;
+use serde::Deserialize;
+use std::collections::BTreeSet;
+#[derive(Deserialize, Clone, Debug, PartialOrd, Ord, PartialEq, Eq, Copy)]
+#[serde(rename_all = "kebab-case")]
+pub enum LanguageFeature {
+    /// This language feature enables experimental syntax that will likely be stabilized in the next major version.
+    /// It may include removals of outdated syntax and introductions of new syntax.
+    V2PreviewSyntax,
+}
+
+impl LanguageFeature {
+    pub fn try_parse(s: &str) -> Result<Self, UnrecognizedLanguageFeature> {
+        match s {
+            "v2-preview-syntax" => Ok(LanguageFeature::V2PreviewSyntax),
+            _ => Err(UnrecognizedLanguageFeature::msg(format!(
+                "Unrecognized language feature: {s}"
+            ))),
+        }
+    }
+}
+
+#[derive(Debug, Clone, Deserialize, Default)]
+pub struct LanguageFeatures(BTreeSet<LanguageFeature>);
+pub type LanguageFeatureIncompatibility = miette::ErrReport;
+pub type UnrecognizedLanguageFeature = miette::ErrReport;
+
+impl LanguageFeatures {
+    /// Checks that the current set of language features is compatible and well-formed.
+    /// Returns a descriptive error message if not.
+    pub fn check_compatibility(&self) -> Result<(), LanguageFeatureIncompatibility> {
+        // we currently only have one language feature, so no checking is required.
+        Ok(())
+    }
+
+    #[must_use]
+    pub fn none() -> Self {
+        Self(BTreeSet::default())
+    }
+
+    pub fn merge(&mut self, other: impl Into<BTreeSet<LanguageFeature>>) {
+        self.0.append(&mut other.into());
+    }
+
+    #[must_use]
+    pub fn contains(&self, feat: LanguageFeature) -> bool {
+        self.0.contains(&feat)
+    }
+}
+impl From<LanguageFeatures> for BTreeSet<LanguageFeature> {
+    fn from(val: LanguageFeatures) -> Self {
+        val.0
+    }
+}
+impl From<BTreeSet<LanguageFeature>> for LanguageFeatures {
+    fn from(val: BTreeSet<LanguageFeature>) -> Self {
+        LanguageFeatures(val)
+    }
+}
+
+impl ValueEnum for LanguageFeature {
```

Comment:
When converting to bitflags as suggested above, I think it would be worth dropping the dependency on `ValueEnum` and clap and just implementing a try-conversion from `String` similar to what we did for target profiles (see compiler/qsc/src/target.rs line 33) along with a matching `to_str` implementation. It's a little more manual to keep in sync, but that shouldn't be a concern until we have very many values at which point we can revisit.